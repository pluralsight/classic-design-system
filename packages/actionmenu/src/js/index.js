import * as vars from '../vars/index.js'

export function calcNestedMenuPosition(parent, menu, origin) {
  return {
    topLeft: {
      left: -parent.width,
      top: `calc(-1 * ${vars.style.menuPaddingVert})`
    },
    topRight: {
      right: parent.width,
      top: `calc(-1 * ${vars.style.menuPaddingVert})`
    },
    bottomRight: {
      right: parent.width,
      top: -Math.abs(parent.height - menu.height)
    },
    bottomLeft: {
      left: -parent.width,
      top: -Math.abs(parent.height - menu.height)
    }
  }[origin]
}
