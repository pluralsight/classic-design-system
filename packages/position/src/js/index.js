const defaultOptions = { bufferWidth: 12 }

export function above(target) {
  if (!target) throw new TypeError('target element required')

  return {
    styleFor(el, options) {
      if (!el) return

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset +
        targetRect.left +
        targetRect.width / 2 -
        elRect.width / 2
      const y =
        window.pageYOffset + targetRect.top - elRect.height - opts.bufferWidth
      return formatOutputAsStyles(x, y)
    }
  }
}

export function aboveLeft(target) {
  if (!target) throw new TypeError('target element required')

  return {
    styleFor(el, options) {
      if (!el) return

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x = window.pageXOffset + targetRect.left
      const y =
        window.pageYOffset + targetRect.top - elRect.height - opts.bufferWidth

      return formatOutputAsStyles(x, y)
    }
  }
}

export function aboveRight(target) {
  if (!target) throw new TypeError('target element required')

  return {
    styleFor(el, options) {
      if (!el) return

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset + targetRect.left + targetRect.width - elRect.width
      const y =
        window.pageYOffset + targetRect.top - elRect.height - opts.bufferWidth

      return formatOutputAsStyles(x, y)
    }
  }
}

export function rightOf(target) {
  if (!target) throw new TypeError('target element required')

  return {
    styleFor(el, options) {
      if (!el) return

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset +
        targetRect.left +
        targetRect.width +
        opts.bufferWidth
      const y =
        window.pageYOffset +
        targetRect.top +
        targetRect.height / 2 -
        elRect.height / 2
      return formatOutputAsStyles(x, y)
    }
  }
}

export function below(target) {
  if (!target) throw new TypeError('target element required')

  return {
    styleFor(el, options) {
      if (!el) return

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset +
        targetRect.left +
        targetRect.width / 2 -
        elRect.width / 2
      const y =
        window.pageYOffset +
        targetRect.top +
        targetRect.height +
        opts.bufferWidth

      return formatOutputAsStyles(x, y)
    }
  }
}

export function belowLeft(target) {
  if (!target) throw new TypeError('target element required')

  return {
    styleFor(el, options) {
      if (!el) return

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const x = window.pageXOffset + targetRect.left
      const y =
        window.pageYOffset +
        targetRect.top +
        targetRect.height +
        opts.bufferWidth

      return formatOutputAsStyles(x, y)
    }
  }
}

export function belowRight(target) {
  if (!target) throw new TypeError('target element required')

  return {
    styleFor(el, options) {
      if (!el) return

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset + targetRect.left + targetRect.width - elRect.width
      const y =
        window.pageYOffset +
        targetRect.top +
        targetRect.height +
        opts.bufferWidth

      return formatOutputAsStyles(x, y)
    }
  }
}

export function leftOf(target) {
  if (!target) throw new TypeError('target element required')

  return {
    styleFor(el, options) {
      if (!el) return

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset + targetRect.left - elRect.width - opts.bufferWidth
      const y =
        window.pageYOffset +
        targetRect.top +
        targetRect.height / 2 -
        elRect.height / 2
      return formatOutputAsStyles(x, y)
    }
  }
}

function overrideDefaultOpts(opts) {
  return { ...opts, ...defaultOptions }
}

function formatOutputAsStyles(x, y) {
  return {
    position: 'absolute',
    left: x + 'px',
    top: y + 'px'
  }
}
