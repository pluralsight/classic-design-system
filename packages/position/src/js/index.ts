interface StyleOptions {
  bufferWidth: number
}
const defaultOptions: StyleOptions = { bufferWidth: 12 }

interface ClipAdjusterProps {
  x: number
  y: number
  elRect: DOMRect
  targetRect: DOMRect
  opts: StyleOptions
}
type ClipAdjustment = [number, number]
type ClipAdjuster = (
  props: ClipAdjusterProps,
  clipY?: boolean
) => ClipAdjustment
const clipAdjust: {
  above: ClipAdjuster
  below: ClipAdjuster
  leftOf: ClipAdjuster
  rightOf: ClipAdjuster
} = {
  above: ({ x, y, elRect, targetRect, opts }, clipY) => [
    [0, x, window.innerWidth - elRect.width].sort((a, b) => a - b)[1],
    y < 0 && !!clipY ? targetRect.bottom + opts.bufferWidth : y
  ],
  below: ({ x, y, elRect, targetRect, opts }, clipY) => [
    [0, x, window.innerWidth - elRect.width].sort((a, b) => a - b)[1],
    y > window.innerHeight + window.pageYOffset && !!clipY
      ? targetRect.top - (elRect.height + opts.bufferWidth)
      : y
  ],
  leftOf: ({ x, y, elRect, targetRect, opts }, clipY) => [
    x < 0 ? targetRect.right + opts.bufferWidth : x,
    [
      0,
      y,
      clipY ? window.innerHeight + window.pageYOffset - elRect.height : y
    ].sort((a, b) => a - b)[1]
  ],
  rightOf: ({ x, y, elRect, targetRect, opts }, clipY) => [
    x > window.innerWidth
      ? targetRect.left - elRect.width - opts.bufferWidth
      : x,
    [
      0,
      y,
      clipY ? window.innerHeight + window.pageYOffset - elRect.height : y
    ].sort((a, b) => a - b)[1]
  ]
}

export type PositionFunction = (
  target: HTMLElement,
  clipY?: boolean
) => {
  styleFor: (
    el: HTMLElement | undefined,
    options?: Partial<StyleOptions>
  ) => PositionStyle
}

export const above: PositionFunction = (target, clipY) => {
  return {
    styleFor(
      el: HTMLElement | undefined,
      options?: Partial<StyleOptions>
    ): PositionStyle {
      if (!el) return {}

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset +
        targetRect.left +
        targetRect.width / 2 -
        elRect.width / 2
      const y =
        window.pageYOffset + targetRect.top - elRect.height - opts.bufferWidth
      return formatOutputAsStyles(
        ...clipAdjust.above({ x, y, targetRect, elRect, opts }, clipY)
      )
    }
  }
}

export const aboveLeft: PositionFunction = (target, clipY) => {
  return {
    styleFor(
      el: HTMLElement | undefined,
      options?: Partial<StyleOptions>
    ): PositionStyle {
      if (!el) return {}

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x = window.pageXOffset + targetRect.left
      const y =
        window.pageYOffset + targetRect.top - elRect.height - opts.bufferWidth

      return formatOutputAsStyles(
        ...clipAdjust.above({ x, y, targetRect, elRect, opts }, clipY)
      )
    }
  }
}

export const aboveRight: PositionFunction = (target, clipY) => {
  return {
    styleFor(
      el: HTMLElement | undefined,
      options?: Partial<StyleOptions>
    ): PositionStyle {
      if (!el) return {}

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset + targetRect.left + targetRect.width - elRect.width
      const y =
        window.pageYOffset + targetRect.top - elRect.height - opts.bufferWidth

      return formatOutputAsStyles(
        ...clipAdjust.above({ x, y, targetRect, elRect, opts }, clipY)
      )
    }
  }
}

export const rightOf: PositionFunction = (target, clipY) => {
  return {
    styleFor(
      el: HTMLElement | undefined,
      options?: Partial<StyleOptions>
    ): PositionStyle {
      if (!el) return {}

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset +
        targetRect.left +
        targetRect.width +
        opts.bufferWidth
      const y =
        window.pageYOffset +
        targetRect.top +
        targetRect.height / 2 -
        elRect.height / 2
      return formatOutputAsStyles(
        ...clipAdjust.rightOf({ x, y, targetRect, elRect, opts }, clipY)
      )
    }
  }
}

export const below: PositionFunction = (target, clipY) => {
  return {
    styleFor(
      el: HTMLElement | undefined,
      options?: Partial<StyleOptions>
    ): PositionStyle {
      if (!el) return {}

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset +
        targetRect.left +
        targetRect.width / 2 -
        elRect.width / 2
      const y =
        window.pageYOffset +
        targetRect.top +
        targetRect.height +
        opts.bufferWidth

      return formatOutputAsStyles(
        ...clipAdjust.below({ x, y, targetRect, elRect, opts }, clipY)
      )
    }
  }
}

export const belowLeft: PositionFunction = (target, clipY) => {
  return {
    styleFor(
      el: HTMLElement | undefined,
      options?: Partial<StyleOptions>
    ): PositionStyle {
      if (!el) return {}

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x = window.pageXOffset + targetRect.left
      const y =
        window.pageYOffset +
        targetRect.top +
        targetRect.height +
        opts.bufferWidth

      return formatOutputAsStyles(
        ...clipAdjust.below({ x, y, targetRect, elRect, opts }, clipY)
      )
    }
  }
}

export const belowRight: PositionFunction = (target, clipY) => {
  return {
    styleFor(
      el: HTMLElement | undefined,
      options?: Partial<StyleOptions>
    ): PositionStyle {
      if (!el) return {}

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset + targetRect.left + targetRect.width - elRect.width
      const y =
        window.pageYOffset +
        targetRect.top +
        targetRect.height +
        opts.bufferWidth

      return formatOutputAsStyles(
        ...clipAdjust.below({ x, y, targetRect, elRect, opts }, clipY)
      )
    }
  }
}

export const leftOf: PositionFunction = (target, clipY) => {
  if (!target) throw new TypeError('target element required')

  return {
    styleFor(
      el: HTMLElement | undefined,
      options?: Partial<StyleOptions>
    ): PositionStyle {
      if (!el) return {}

      const opts = overrideDefaultOpts(options)
      const targetRect = target.getBoundingClientRect()
      const elRect = el.getBoundingClientRect()
      const x =
        window.pageXOffset + targetRect.left - elRect.width - opts.bufferWidth
      const y =
        window.pageYOffset +
        targetRect.top +
        targetRect.height / 2 -
        elRect.height / 2
      return formatOutputAsStyles(
        ...clipAdjust.leftOf({ x, y, targetRect, elRect, opts }, clipY)
      )
    }
  }
}

function overrideDefaultOpts(
  opts: Partial<StyleOptions> | undefined
): StyleOptions {
  return { ...opts, ...defaultOptions }
}

export type PositionStyle = Pick<
  React.CSSProperties,
  'position' | 'left' | 'top'
>
function formatOutputAsStyles(x: number, y: number): PositionStyle {
  return {
    position: 'absolute',
    left: x.toString() + 'px',
    top: y.toString() + 'px'
  }
}
