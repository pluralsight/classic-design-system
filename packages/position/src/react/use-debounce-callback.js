import React from 'react'

export default function useDebounceCallback(fn, delay = 100) {
  const timeout = React.useRef()

  const debounced = React.useCallback(
    function () {
      const self = this
      const args = arguments

      function later() {
        timeout.current = null
        fn.apply(self, args)
      }

      clearTimeout(timeout.current)
      timeout.current = setTimeout(later, delay)
    },
    [fn, delay]
  )

  React.useEffect(
    () => () => {
      if (!timeout.current) return
      clearTimeout(timeout.current)
      timeout.current = null
    },
    [fn, delay]
  )

  return debounced
}
