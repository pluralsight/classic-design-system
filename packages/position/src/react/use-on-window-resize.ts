import React from 'react'

import { useDebounceCallback } from '@pluralsight/ps-design-system-util'

export default function useOnWindowResize(handler: (evt: Event) => void) {
  const handleScroll = useDebounceCallback(handler, 10)

  React.useEffect(
    function bindResizeEvent() {
      window.addEventListener('resize', handleScroll)

      return function unbind() {
        window.removeEventListener('resize', handleScroll)
      }
    },
    [handleScroll]
  )
}
