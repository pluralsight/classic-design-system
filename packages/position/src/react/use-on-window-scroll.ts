import React from 'react'

import { useDebounceCallback } from '@pluralsight/ps-design-system-util'

export default function useOnWindowScroll(handler: (evt: Event) => void) {
  const handleScroll = useDebounceCallback(handler, 10)

  React.useEffect(
    function bindScrollEvent() {
      window.addEventListener('scroll', handleScroll, true)

      return function unbind() {
        window.removeEventListener('scroll', handleScroll)
      }
    },
    [handleScroll]
  )
}
