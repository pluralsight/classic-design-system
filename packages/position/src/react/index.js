import PropTypes from 'prop-types'
import React from 'react'
import { canUseDOM } from 'exenv'
import {
  cloneElementWithRef,
  createUniversalPortal
} from '@pluralsight/ps-design-system-util'

import * as positionFns from '../js/index.js'

import useOnWindowResize from './use-on-window-resize.js'
import useOnWindowScroll from './use-on-window-scroll.js'

export const Above = React.forwardRef((props, ref) => {
  return <Position position={positionFns.above} ref={ref} {...props} />
})
Above.displayName = 'Above'

export const AboveLeft = React.forwardRef((props, ref) => {
  return <Position position={positionFns.aboveLeft} ref={ref} {...props} />
})
AboveLeft.displayName = 'AboveLeft'

export const AboveRight = React.forwardRef((props, ref) => {
  return <Position position={positionFns.aboveRight} ref={ref} {...props} />
})
AboveRight.displayName = 'AboveRight'

export const Below = React.forwardRef((props, ref) => {
  return <Position position={positionFns.below} ref={ref} {...props} />
})
Below.displayName = 'Below'

export const BelowLeft = React.forwardRef((props, ref) => {
  return <Position position={positionFns.belowLeft} ref={ref} {...props} />
})
BelowLeft.displayName = 'BelowLeft'

export const BelowRight = React.forwardRef((props, ref) => {
  return <Position position={positionFns.belowRight} ref={ref} {...props} />
})
BelowRight.displayName = 'BelowRight'

export const RightOf = React.forwardRef((props, ref) => {
  return <Position position={positionFns.rightOf} ref={ref} {...props} />
})
RightOf.displayName = 'RightOf'

export const LeftOf = React.forwardRef((props, ref) => {
  return <Position position={positionFns.leftOf} ref={ref} {...props} />
})
LeftOf.displayName = 'LeftOf'

const Position = React.forwardRef((props, forwardedRef) => {
  const { target, position: positionFn, when } = props
  const [shownOnce, setShownOnce] = React.useState(false)
  const [style, setStyle] = React.useState({ position: 'absolute' })

  const ref = React.useRef()
  React.useImperativeHandle(forwardedRef, () => ref.current)

  const child = React.Children.only(props.children)
  const showRef = React.useRef()
  const showEl = React.cloneElement(props.show, {
    ref: showRef,
    style: { ...props.show.props.style, ...style }
  })

  const updateStyle = React.useCallback(() => {
    const targetNode = target ? target.current : ref.current
    if (!showRef.current || !targetNode) return

    const nextStyle = positionFn(targetNode).styleFor(showRef.current)
    setStyle(nextStyle)
    setShownOnce(true)
  }, [positionFn, target])

  useOnWindowResize(evt => updateStyle())
  useOnWindowScroll(evt => {
    const isInner = !showRef.current || showRef.current.contains(evt.target)
    if (isInner) return

    updateStyle()
  })
  React.useEffect(() => {
    if (!when) return
    const timerId = delayUntilNextTick(() => updateStyle())

    return () => clearTimeout(timerId)
  }, [updateStyle, when])

  return (
    <>
      {target ? child : cloneElementWithRef(child, ref, {})}

      {createUniversalPortal(
        <div style={{ visibility: shownOnce ? 'visible' : 'hidden' }}>
          {when && showEl}
        </div>,
        props.inNode
      )}
    </>
  )
})

Position.propTypes = {
  children: PropTypes.element.isRequired,
  inNode: PropTypes.any, // NOTE: really HTML Element, but SSR not happy when mentioned.
  position: PropTypes.func.isRequired,
  show: PropTypes.element,
  target: PropTypes.any, // NOTE: really HTML Element, but SSR not happy when mentioned.
  when: PropTypes.bool
}

Position.defaultProps = {
  inNode: canUseDOM ? document.body : null,
  when: true
}

function delayUntilNextTick(fn) {
  // eslint-disable-next-line
  return setTimeout(fn, 1)
}
