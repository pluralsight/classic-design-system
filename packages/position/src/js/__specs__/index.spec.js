import { render, waitForElement } from '@testing-library/react'
import React from 'react'

import * as positionFns from '../index.js'

describe('#below', () => {
  const { below } = positionFns

  it('throws if no target el given', () => {
    expect(_ => below()).toThrow(/target element required/)
  })

  it('returns undefined if no el given', async () => {
    const { getByTestId } = render(<div data-testid="target-el" />)
    const targetEl = await waitForElement(() => getByTestId('target-el'))
    expect(below(targetEl).styleFor()).toBeUndefined()
  })

  it('returns styles', async () => {
    const targetEl = await createEl()
    const el = await createEl({
      width: 100,
      height: 100
    })

    expect(below(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '12px',
      left: '0px'
    })
  })

  it('adjusts for target dimensions', async () => {
    const targetEl = await createEl({
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 100,
      height: 100
    })

    expect(below(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '212px',
      left: '50px'
    })
  })

  it('adjusts for target placement', async () => {
    const targetEl = await createEl({
      height: 200,
      width: 200,
      top: 50,
      left: 25
    })
    const el = await createEl({
      width: 100,
      height: 100
    })

    expect(below(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '262px',
      left: '75px'
    })
  })

  it('adjusts for element width', async () => {
    const targetEl = await createEl({
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 200,
      height: 100
    })

    expect(below(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '212px',
      left: '0px'
    })
  })
})

describe('#belowLeft', () => {
  const { belowLeft } = positionFns

  it('throws if no target el given', () => {
    expect(_ => belowLeft()).toThrow(/target element required/)
  })

  it('returns undefined if no el given', async () => {
    const { getByTestId } = render(<div data-testid="target-el" />)
    const targetEl = await waitForElement(() => getByTestId('target-el'))
    expect(belowLeft(targetEl).styleFor()).toBeUndefined()
  })

  it('returns styles', async () => {
    const targetEl = await createEl()
    const el = await createEl({
      width: 100,
      height: 100
    })

    expect(belowLeft(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '12px',
      left: '0px'
    })
  })

  it('adjusts for target dimensions', async () => {
    const targetEl = await createEl({
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 100,
      height: 100
    })

    expect(belowLeft(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '212px',
      left: '0px'
    })
  })

  it('adjusts for target placement', async () => {
    const targetEl = await createEl({
      height: 200,
      width: 200,
      top: 50,
      left: 25
    })
    const el = await createEl({
      width: 100,
      height: 100
    })

    expect(belowLeft(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '262px',
      left: '25px'
    })
  })

  it('adjusts for element width', async () => {
    const targetEl = await createEl({
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 200,
      height: 100
    })

    expect(belowLeft(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '212px',
      left: '0px'
    })
  })
})

describe('#belowRight', () => {
  const { belowRight } = positionFns

  it('throws if no target el given', () => {
    expect(_ => belowRight()).toThrow(/target element required/)
  })

  it('returns undefined if no el given', async () => {
    const { getByTestId } = render(<div data-testid="target-el" />)
    const targetEl = await waitForElement(() => getByTestId('target-el'))
    expect(belowRight(targetEl).styleFor()).toBeUndefined()
  })

  it('returns styles', async () => {
    const targetEl = await createEl()
    const el = await createEl({ width: 150, height: 100 })

    expect(belowRight(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '12px',
      left: '0px'
    })
  })

  it('adjusts for target dimensions', async () => {
    const targetEl = await createEl({
      height: 200,
      width: 100
    })
    const el = await createEl({
      width: 150,
      height: 100
    })

    expect(belowRight(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '212px',
      left: '0px'
    })
  })

  it('adjusts for target placement', async () => {
    const targetEl = await createEl({
      height: 200,
      width: 150,
      top: 50,
      left: 125
    })
    const el = await createEl({
      width: 150,
      height: 100
    })

    expect(belowRight(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '262px',
      left: '125px'
    })
  })

  it('adjusts for element width', async () => {
    const targetEl = await createEl({
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 200,
      height: 100
    })

    expect(belowRight(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '212px',
      left: '0px'
    })
  })
})

describe('#above', () => {
  const { above } = positionFns

  it('returns undefined if no el given', async () => {
    const { getByTestId } = render(<div data-testid="target-el" />)
    const targetEl = await waitForElement(() => getByTestId('target-el'))
    expect(above(targetEl).styleFor()).toBeUndefined()
  })

  it('adjusts for target placement and dimensions', async () => {
    const targetEl = await createEl({
      top: 100,
      left: 25,
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 100,
      height: 50
    })

    expect(above(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '38px',
      left: '75px'
    })
  })

  it('adjusts for element width', async () => {
    const targetEl = await createEl({
      top: 100,
      left: 25,
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 300,
      height: 50
    })

    expect(above(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '38px',
      left: '0px'
    })
  })
})

describe('#aboveLeft', () => {
  const { aboveLeft } = positionFns

  it('returns undefined if no el given', async () => {
    const { getByTestId } = render(<div data-testid="target-el" />)
    const targetEl = await waitForElement(() => getByTestId('target-el'))
    expect(aboveLeft(targetEl).styleFor()).toBeUndefined()
  })

  it('adjusts for target placement and dimensions', async () => {
    const targetEl = await createEl({
      top: 100,
      left: 25,
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 100,
      height: 50
    })

    expect(aboveLeft(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '38px',
      left: '25px'
    })
  })

  it('adjusts for element width', async () => {
    const targetEl = await createEl({
      top: 100,
      left: 25,
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 300,
      height: 50
    })

    expect(aboveLeft(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '38px',
      left: '25px'
    })
  })
})

describe('#aboveRight', () => {
  const { aboveRight } = positionFns

  it('returns undefined if no el given', async () => {
    const { getByTestId } = render(<div data-testid="target-el" />)
    const targetEl = await waitForElement(() => getByTestId('target-el'))
    expect(aboveRight(targetEl).styleFor()).toBeUndefined()
  })

  it('adjusts for target placement and dimensions', async () => {
    const targetEl = await createEl({
      top: 100,
      left: 25,
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 100,
      height: 50
    })

    expect(aboveRight(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '38px',
      left: '125px'
    })
  })

  it('adjusts for element width', async () => {
    const targetEl = await createEl({
      top: 100,
      left: 25,
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 300,
      height: 50
    })

    expect(aboveRight(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '38px',
      left: '0px'
    })
  })
})

describe('#rightOf', () => {
  const { rightOf } = positionFns

  it('returns undefined if no el given', async () => {
    const { getByTestId } = render(<div data-testid="target-el" />)
    const targetEl = await waitForElement(() => getByTestId('target-el'))
    expect(rightOf(targetEl).styleFor()).toBeUndefined()
  })

  it('adjusts for target dimensions and placement', async () => {
    const targetEl = await createEl({
      top: 50,
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 100,
      height: 50
    })

    expect(rightOf(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '125px',
      left: '212px'
    })
  })

  it('adjusts for element height', async () => {
    const targetEl = await createEl({
      top: 50,
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 100,
      height: 250
    })

    expect(rightOf(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '25px',
      left: '212px'
    })
  })
})

describe('#leftOf', () => {
  const { leftOf } = positionFns

  it('returns undefined if no el given', async () => {
    const { getByTestId } = render(<div data-testid="target-el" />)
    const targetEl = await waitForElement(() => getByTestId('target-el'))
    expect(leftOf(targetEl).styleFor()).toBeUndefined()
  })

  it('adjusts for target dimensions and placement', async () => {
    const targetEl = await createEl({
      top: 50,
      left: 300,
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 100,
      height: 50
    })

    expect(leftOf(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '125px',
      left: '188px'
    })
  })

  it('adjusts for element width', async () => {
    const targetEl = await createEl({
      top: 50,
      left: 300,
      height: 200,
      width: 200
    })
    const el = await createEl({
      width: 25,
      height: 50
    })

    expect(leftOf(targetEl).styleFor(el)).toEqual({
      position: 'absolute',
      top: '125px',
      left: '263px'
    })
  })
})

function randStr() {
  return (
    Math.random()
      .toString(36)
      .substring(2, 15) +
    Math.random()
      .toString(36)
      .substring(2, 15)
  )
}

async function createEl(boundingClientRectOpts) {
  const id = randStr()
  const { getByTestId } = render(<div data-testid={id} />)
  const el = await waitForElement(() => getByTestId(id))
  el.getBoundingClientRect = () => ({
    width: 0,
    height: 0,
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    ...boundingClientRectOpts
  })
  return el
}
