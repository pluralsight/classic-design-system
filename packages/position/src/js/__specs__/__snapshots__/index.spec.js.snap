// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#edgeCases #above: bottomLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #above: bottomRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #above: topLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #above: topRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #aboveLeft: bottomLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #aboveLeft: bottomRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #aboveLeft: topLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #aboveLeft: topRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #aboveRight: bottomLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #aboveRight: bottomRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #aboveRight: topLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #aboveRight: topRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #below: bottomLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #below: bottomRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #below: topLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #below: topRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #belowLeft: bottomLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #belowLeft: bottomRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #belowLeft: topLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #belowLeft: topRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #belowRight: bottomLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #belowRight: bottomRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "372px",
}
`;

exports[`#edgeCases #belowRight: topLeft 1`] = `
Object {
  "left": "0px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #belowRight: topRight 1`] = `
Object {
  "left": "400px",
  "position": "absolute",
  "top": "28px",
}
`;

exports[`#edgeCases #leftOf: bottomLeft 1`] = `
Object {
  "left": "28px",
  "position": "absolute",
  "top": "400px",
}
`;

exports[`#edgeCases #leftOf: bottomRight 1`] = `
Object {
  "left": "372px",
  "position": "absolute",
  "top": "400px",
}
`;

exports[`#edgeCases #leftOf: topLeft 1`] = `
Object {
  "left": "28px",
  "position": "absolute",
  "top": "0px",
}
`;

exports[`#edgeCases #leftOf: topRight 1`] = `
Object {
  "left": "372px",
  "position": "absolute",
  "top": "0px",
}
`;

exports[`#edgeCases #rightOf: bottomLeft 1`] = `
Object {
  "left": "28px",
  "position": "absolute",
  "top": "400px",
}
`;

exports[`#edgeCases #rightOf: bottomRight 1`] = `
Object {
  "left": "372px",
  "position": "absolute",
  "top": "400px",
}
`;

exports[`#edgeCases #rightOf: topLeft 1`] = `
Object {
  "left": "28px",
  "position": "absolute",
  "top": "0px",
}
`;

exports[`#edgeCases #rightOf: topRight 1`] = `
Object {
  "left": "372px",
  "position": "absolute",
  "top": "0px",
}
`;
